src/stream/src/executor/backfill.rs

```diff
@@ -0,0 +1,400 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::mem;
+use std::ops::Bound;
+use std::sync::Arc;
+
+use async_stack_trace::StackTrace;
+use either::Either;
+use futures::stream::select_with_strategy;
+use futures::{pin_mut, stream, StreamExt};
+use futures_async_stream::try_stream;
+use risingwave_common::array::{Op, Row, StreamChunk};
+use risingwave_common::buffer::BitmapBuilder;
+use risingwave_common::catalog::Schema;
+use risingwave_hummock_sdk::HummockReadEpoch;
+use risingwave_storage::table::batch_table::storage_table::StorageTable;
+use risingwave_storage::table::TableIter;
+use risingwave_storage::StateStore;
+
+use super::error::StreamExecutorError;
+use super::{expect_first_barrier, BoxedExecutor, Executor, ExecutorInfo, Message};
+use crate::executor::PkIndices;
+use crate::task::{ActorId, CreateMviewProgress};
+
+/// An implementation of the RFC: Use Backfill To Let Mv On Mv Stream Again.(https://github.com/risingwavelabs/rfcs/pull/13)
+/// `BackfillExecutor` is used to create a materialized view on another materialized view.
+///
+/// It can only buffer chunks between two (checkpoint) barriers instead of unbundled memory usage of
+/// `RearrangedChainExecutor`.
+///
+/// It uses the latest epoch to read the snapshot of the upstream mv during two barriers and all the
+/// `StreamChunk` of the snapshot read will forward to the downstream.
+///
+/// It uses `current_pos` to record the progress of the backfill (the pk of the upstream mv) and
+/// `current_pos` is initiated as an empty `Row`.
+///
+/// All upstream messages during the two barriers interval will be buffered and decide to forward or
+/// ignore based on the `current_pos` at the end of the later barrier. Once `current_pos` reaches
+/// the end of the upstream mv pk, the backfill would finish.
+///
+/// Notice: The pk we talk about here refers to the storage primary key.
+pub struct BackfillExecutor<S: StateStore> {
+    // Upstream table
```

Comment:
ditto for other occurances

```suggestion
    /// Upstream table
```